msal-obo-sample/src/main/java/com/microsoft/azure/msalobosample/OboAuthProvider.java

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.microsoft.azure.msalobosample;
+
+import com.google.common.hash.Hashing;
+import com.microsoft.aad.msal4j.ClientCredentialFactory;
+import com.microsoft.aad.msal4j.ConfidentialClientApplication;
+import com.microsoft.aad.msal4j.IAuthenticationResult;
+import com.microsoft.aad.msal4j.MsalException;
+import com.microsoft.aad.msal4j.OnBehalfOfParameters;
+import com.microsoft.aad.msal4j.SilentParameters;
+import com.microsoft.aad.msal4j.UserAssertion;
+import com.microsoft.graph.authentication.BaseAuthenticationProvider;
+import org.jetbrains.annotations.NotNull;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.cache.CacheManager;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.oauth2.provider.authentication.OAuth2AuthenticationDetails;
+import org.springframework.stereotype.Component;
+
+import javax.annotation.Nonnull;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import java.util.concurrent.CompletableFuture;
+
+@Component
+class OboAuthProvider extends BaseAuthenticationProvider {
+
+    @Value("${security.oauth2.client.authority}")
+    private String authority;
+
+    @Value("${security.oauth2.client.client-id}")
+    private String clientId;
+
+    @Value("${security.oauth2.client.client-secret}")
+    private String secret;
+
+    @Value("${aad.graphDefaultScope}")
+    private String scope;
+
+    @Autowired
+    CacheManager cacheManager;
+
+    @NotNull
+    @Override
+    public CompletableFuture<String> getAuthorizationTokenAsync(@Nonnull URL url) {
+
+        // Gets incoming access token and generates cache key. The cache key will be used to store
+        // the tokens for the incoming request.
+        String authToken = this.getAuthToken();
+        String cacheKey = Hashing.sha256().hashString(authToken, StandardCharsets.UTF_8).toString();
+
+        IAuthenticationResult authResult;
+        ConfidentialClientApplication application = null;
+        try {
+            application = ConfidentialClientApplication
+                    .builder(clientId, ClientCredentialFactory.createFromSecret(secret))
+                    .authority(authority)
+                    .build();
+
+            String cachedTokens = cacheManager.getCache("tokens").get(cacheKey, String.class);
+            if (cachedTokens != null) {
+                application.tokenCache().deserialize(cachedTokens);
+            }
+
+            SilentParameters silentParameters =
```

Comment:
should we switch to new API with no silent call ?

Reply:
Yep, was waiting for this to be released in MSAL. Now that it has I'll make the updates.